src/main/scala/com/databricks/spark/sql/perf/mllib/ReflectionUtils.scala

```diff
@@ -0,0 +1,35 @@
+package com.databricks.spark.sql.perf.mllib
+
+import scala.reflect.ClassTag
+import scala.reflect.runtime.universe._
+
+/** Exposes methods to simplify implementation of classes like MLParameters. */
+private[perf] object ReflectionUtils {
+
+  /**
+   * Given an instance [[obj]] of a class whose constructor arguments are all of type Option[Any],
+   * returns a map of key-value pairs (argName -> argValue) where argName is the name
+   * of a constructor argument with a defined (not None) value and argValue is the corresponding
+   * value.
+   */
+  def getConstructorArgs[T: TypeTag: ClassTag](obj: T): Map[String, Any] = {
```

Comment:
This method allows me to avoid a bunch of statements of the form
```
if (parameter.isDefined) {
  map ++= (parameter -> numFeatures.get.toString)
}
```
in `MLParameters.toMap` -- let me know if it'd be better to not use reflection/do this another way.

Reply:
Hmm... interesting, use reflection here, can we also use reflection to generate the `MLParameters.copy` method ? So we can avoid duplicate code.

Reply:
I tried to do this but it got messy so I just stuck with writing out the copy constructor :P